### ISSUE TITLE
Parsing Disgaea Savegames

### ISSUE THREAD

**Issue by mechpaul** [Xkeeper0/disgaea-pc-tools#1]
Hello. I have hooked up a debugger to the Disgaea exe and I'm watching it unpack a save file. I am writing here what it does to the file.

DISGAEA SAVE FILE:
0x00 - 0x1F = You can consider these bytes to be constant. The save game checker validates that these bytes are not changed. If they are changed, it does some weird voodoo magic that I haven't deciphered yet. But these bytes might as well be static.
0x20 - UINT = This is the xorkey as you properly assumed
0x24 - 0x27 = unknown right now
0x28 - uint32_t = (bytes << 2) == length compressed file
0x3B - uint32_t = (bytes << 2) == length decompressed file

Run xorkey over the file as you're doing. Now you start the decompression procedure.

assert(header == YKCMP_V1)
version = (uint32_t) ykcmp[8]
assert(version == 0x04)
lengthCurrentObject = (uint32_t) ykcmp[12]
lengthDecompressedObject = (uint32_t) ykcmp[16]

From here you start unpacking. You've got the algorithm correct, but the code in the assembly is much cleaner.

```
byte = getByte()
if (byte & 0x80) == 0:
    length = byte;
    lookback = 0;
    memcpy(compressedObject, decompressedObject, byte);
    compressedObject += byte;
    decompressedObject += byte;
    continue;
elif (byte & 0x40) == 0:
    length = ((byte >> 4) & 3) + 1
    lookback = length;
elif (byte & 0x20) == 0:
    nextByte = getByte();
    length = (byte & 0x1F) + 2;
    lookback = nextByte;
else:
    #blah math
memcpy(compressedObject - lookback - 1, compressedObject, length);
compressedObject += length;
decompressedObject += length;
```

For the first part of the unpacked file (where we can start reading character data), it uses the first name in the list of characters as your main character. Then it fetches the first character level (abs offset 0x30) and gets the playtime. 

```
FILE* f = fopen("SAVE000.DAT.unp", "rb");
fseek(f, 8, SEEK_SET);
unsigned short hours;
unsigned char minutes;
unsigned char seconds;
fread(&hours, sizeof(hours), 1, f);
fread(&minutes, sizeof(minutes), 1, f);
fread(&seconds, sizeof(seconds), 1, f);
```

That's all I've got at the moment. I'll do some more debugging and figure out the rest of the format later.

**Comment by Xkeeper0** [comment]
Hey, cool to see you here!

I ended up figuring out the "compressed length << 4" bit of the header last night -- turns out that it was the last thing stopping the recreated files from working, though my method was just to copy all known data from a good save file to a test save file, then compare the differences (ended up being about 5 bytes) and updating them one by one until it worked ... then walking back and figuring out which one specifically made it not crash. A little old fashioned, but it worked. 

At some point I need to get set up with some PC debugging tools and actually learn disassembling. I imagine my knowledge of 6502 will help :p

The timer I managed to figure out by just replacing the first X bytes of the save file with `01 02 03 ... 1F`; the timer ended up working out to `0A09:0B:0C` (though in decimal). There are some other important values in the first few bytes as well, since setting them to various values will change which "mode" the game is in (main vs Etna mode, etc) as well as cause some effects to spawn as soon as the file loads.

I guess now that I finally have the important basic parts working, I can start implementing these as editable chunks in the file(s), though. :D

**Comment by mechpaul** [comment]
I am now perturbed.

[WTF #1] I set a breakpoint on the fopen function and created a savegame. My BP was never hit.
[WTF #2] I created BPs on every place where it used the string "SAVE%03d.DAT" and then created a savegame. None of my BPs were hit.
[WTF #3] Disgaea does not use the CreateFileA function.

So, how does Disgaea create a savegame? Right now it's voodoo magic, and I am not a mystic.

**Comment by mechpaul** [comment]
![image](https://cloud.githubusercontent.com/assets/677301/13901163/efc09e4a-edd7-11e5-8058-214b75f6a660.png)

Looks like I'll need to look more into Steam saves to figure out how these work.

**Comment by Xkeeper0** [comment]
The only thing I can really add to this part of the discussion is that I can only get predictable results (e.g. edited files actually being opened) if I **put steam into offline mode**. With it online (even with cloud saves off) it seems to use Steam's caching or something.

It's been a big pain in the ass.

**Comment by Xkeeper0** [comment]
Okay so even offline mode doesn't help if the remotecache file is in the save game folder

`C:\Program Files (x86)\Steam\userdata\XXXXXXXX\405900\`

I was doing some basic save file editing (as in, just opening + saving a file, not changing) and it was giving me a glitched, softlock file whenever I tried to open it. Deleting that file makes it work again.

What the fuck, NIS?

**Comment by mechpaul** [comment]
It's not an NIS problem. Calculate the SHA1 of the file you edited and change the remotecache.vdf file to reflect the change. Steam should accept your change after updating the vdf file. Untested, but likely.

**Comment by Xkeeper0** [comment]
Nah, it has nothing to do with that. Turns out that it's actually `game.cfg` -- for whatever asinine reason that file sometimes dictates that edited save files just don't load properly. Deleting it (and remotecache) ensures that save files work properly every time.

NIS.

**Comment by Xkeeper0** [comment]
I confirmed the above by actually writing something that would update the file with the correct sha1 (and size) when copying over new files. Still didn't work. It seems like the only "real" way to fix it when this happens is to start an entirely new save file (??) and save that, then go back to copying over new ones. Saving itself seems to cause the problem, sometimes deleting the remotecache and gamecfg files fixes it too.

ughhhhhhh
